tests/ProductionConfigTests.js

```diff
@@ -24,28 +24,710 @@ https://github.com/GPII/universal/blob/master/LICENSE.txt
+                            "mag-factor": 1.5,
+                            "screen-position": "full-screen"
+                        },
+                        "http://registry.gpii.net/applications/org.alsa-project": {
+                            "masterVolume": 50
+                        },
+                        "http://registry.gpii.net/common/volume": 0.5,
+                        "http://registry.gpii.net/applications/org.gnome.desktop.interface": {
+                            "cursor-size": 90,
+                            "text-scaling-factor": 0.75
+                        },
+                        "http://registry.gpii.net/common/fontSize": 9,
+                        "http://registry.gpii.net/common/cursorSize": 0.9,
+                        "http://registry.gpii.net/common/backgroundColor": "black",
+                        "http://registry.gpii.net/common/fontFaceFontName": [
+                            "Comic Sans"
+                        ],
+                        "http://registry.gpii.net/common/fontFaceGenericFontFace": "sans serif",
+                        "http://registry.gpii.net/common/magnification": 1.5,
+                        "http://registry.gpii.net/common/adaptationPreference": [
+                            {
+                                "adaptationType": "caption",
+                                "language": "en"
+                            }
+                        ],
+                        "http://registry.gpii.net/common/tableOfContents": false
+                    }
+                }
+            }
+        }
+    },
+    "timestampCreated": new Date().toISOString(),
+    "timestampUpdated": null
+};
+
+gpii.tests.productionConfigTesting.carlaKey = "carla";
+gpii.tests.productionConfigTesting.carlaTokenRequestPayload = {
+    "username": gpii.tests.productionConfigTesting.carlaKey,
+    "gpiiKey": gpii.tests.productionConfigTesting.carlaKey,
+    "password": "dummy",
+    "client_id": "pilot-computer",
+    "client_secret": "pilot-computer-secret",
+    "grant_type": "password"
+};
+
+gpii.tests.productionConfigTesting.device = {
+    OS: {
+        id: "linux"
+    },
+    solutions: [{
+        "id": "org.gnome.desktop.a11y.magnifier"
+    }]
+};
+gpii.tests.productionConfigTesting.prefsUpdate = {
+    "contexts": {
+        "gpii-default": {
+            "name": "Default preferences",
+            "preferences": {
+                "http://registry.gpii.net/common/setting1": 12
+            }
+        }
+    }
+};
+
+fluid.defaults("gpii.tests.productionConfigTesting.testCaseHolder", {
+    gradeNames: ["kettle.test.testCaseHolder"],
+    components: {
+        accessTokenRequest: {
+            type: "kettle.test.request.http",
+            options: {
+                headers: {
+                    "Content-Type": "application/x-www-form-urlencoded"
+                },
+                path: "/access_token",
+                port: 9082,
```

Comment:
Taking the `port` property out of the code works, but adding a `hostname` to the config file doesn't.  The result is a 404.  What is the correct name of the property to add to the config file to specify the host/hostname?

Reply:
What I found in kettle doc is `host` or `hostname`. In GPII config file is `baseUrl` for configuring url data source, which may not be the one for configuring a kettle server.

Reply:
I just tried `host` instead of `hostname` and that worked.  So, I'll go with `host`.

Reply:
My mistake -- putting a `host` in the configuration file does not work either.  I've gone through code and documentation, and I can't see where or how one sets the host/hostname in a configuration file.

Reply:
Also, removing the `port` property from the component request specifications for [the tests defined here](https://github.com/GPII/universal/blob/ebbf7ebcb43179548d8754908b20ac6d2afd6678/tests/ProductionConfigTests.js#L253) works.  The `port` is correctly taken from the configuration file in these cases.

But, removing the port number from the specification [you are referring to](https://github.com/GPII/universal/blob/ebbf7ebcb43179548d8754908b20ac6d2afd6678/tests/ProductionConfigTests.js#L177) does not work.  The `port` used is 8081, the default for a `kettle.test.request.http` component.  I'm still trying to figure out why the difference.

Reply:
After thinking more about this, instead of modifying `host` in the config file that will affect the entire kettle server setup, another approach is to create an add-on grade that uses "distributeOptions" to apply host config to kettle test requests that need to send requests to `flowmanager:9082.

This add-on grade looks like:
```
fluid.defaults("gpii.tests.productionConfigTesting.hostConfig", {
    gradeNames: ["fluid.component"],
    distributeOptions: {
        hostConfig: {
            record: {
                hostname: "flowmanager",
                port: 9082
            },
            target: "{that kettle.test.request.http}.options"
        }
    }
});
```
It can be applied as an add-on grade for this test case holder, such as at line 708 & 731.

Reply:
I tried that, but I'm not sure I did it correctly -- it failed.

What I did:
1. added the `fluid.defaults("gpii.tests.productionConfigTesting.hostConfig", ...` as you suggested, 
2. added that gradeName to the `gpii.tests.productionConfigTesting.testCaseHolder` on [line 168](https://github.com/GPII/universal/blob/ebbf7ebcb43179548d8754908b20ac6d2afd6678/tests/ProductionConfigTests.js#L168) such that it now reads:

```
fluid.defaults("gpii.tests.productionConfigTesting.testCaseHolder", {
    gradeNames: [kettle.test.testCaseHolder, gpii.tests.productionConfigTesting.hostConfig]
    ...
```
The outcome is that the tests for pre-loading Couch with the 'user' keys and prefsSafe fail because the `gpii.tests.productionConfigTesting.hostConfig` is forced on the requests for accessing the data base:
```
...
18:58:50.923:  Sending a POST request to: /gpii on port 5984
18:58:50.925:  jq: FAIL: Module "gpii.tests.untrusted.production.config tests" 
    Test name "Flow manager production tests - add test GPII keys and PrefsSafe" - 
    Message: Error making request to /gpii: connect ECONNREFUSED 172.17.0.4:5984
...
```
If this is the correct way to set things up, then I'll have to come with a way of not using that grade for the couch requests -- somehow turn it on/off as needed.

Reply:
@klown, I issued [a pull request](https://github.com/klown/universal/pull/7) against your branch to address this comment.

While working on `tests/ProductionConfigTests.js`, I felt it can be split into a few files to be more self-contained and easier to work on. The structure in my mind is:

- Create a new tests/production directory. This directory has these files:
1. all-tests.js, the umbrella script run by `scripts/vagrantCloudBasedContainers.sh`. It contains individual tests below
2. CloudStatusProductionTests.js - The part to test /health, /ready etc
3. SettingsGetProductionTests.js - test GET /settings
4. SettingsPutProductionTests.js - test PUT /settings

There could be a utility js file for common utlity functions such as reading data from external files, adding/removing test documents from CouchDB etc.

What do you think?

Reply:
Sounds like plan, but just in case I don't understand and to add some details so I know we are on the same page...

The other two major pieces here are the common test definitions located in the `shared` directory, and the changes to the current test code that runs under a development configuration.

An example of common shared test definitions is the `FlowManagerSettingsGetTestDefs.js` file I created for this pull request.  (Aside:  the current test definitions in `CloudStatusTests.js` similarly need to be factored out and moved to the `shared` directory, for this approach).

The current development configuration test code files (`SettingsGetTests.js`, `SettingsPutTests.js`, and now`CloudStatusTests.js`) stay where they are, but are modified to use the shared test definitions.  I imagine they will become fairly brief.

Does this sound correct to you?

Reply:
Yes, sounds correct to me.